File: python/day01/day01.py
# %%
import numpy as np

test_str = """3   4
4   3
2   5
1   3
3   9
3   3""".splitlines()


def parselines(fname="input.txt"):
    return open(fname, "r").readlines()


def read_into_lists(lines):
    return np.array([[int(x.strip()) for x in line.split()]
                     for line in lines]).T


def part1(lines):
    lists = read_into_lists(lines)

    sorted_lists = np.sort(lists, axis=1)
    return sum(np.abs(sorted_lists[1] - sorted_lists[0]))


def part2(lines):
    lists = read_into_lists(lines)
    
    def count_occurrences(needle, haystack):
        return sum(np.array(haystack) == needle)
    return sum(n*count_occurrences(n, lists[1]) for n in lists[0])


print(part1(test_str))
print(part1(parselines()))
print(part2(test_str))
print(part2(parselines()))

# %%

File: haskell/day01/Day01.hs
import Data.List (sort, transpose)

parseFile :: String -> IO [[Int]]
parseFile fname = do
  let parseLine = map read . words
  contents <- readFile fname
  let allLines = lines contents
  return $ map parseLine allLines

part1 :: [[Int]] -> Int
part1 pairs =
  let lists = transpose pairs
      list0 = sort (lists !! 0)
      list1 = sort (lists !! 1)
   in sum $ map abs $ zipWith (-) list0 list1

part2 :: [[Int]] -> Int
part2 pairs =
  let lists = transpose pairs
      list0 = lists !! 0
      list1 = lists !! 1
      countOccurrences n = length . filter (n ==)
   in sum [n * countOccurrences n list1 | n <- list0]

main :: IO ()
main = do
  pairs <- parseFile "input.txt"
  print $ part1 pairs
  print $ part2 pairs
